[WIP] Add DINO DETR Model to HuggingFace Transformers

konstantinos-p opened the pull request:
## What does this PR do?

This PR introduces the **DINO DETR (DEtection TRansformer with DIstillation)** model (https://arxiv.org/abs/2203.03605) to the Hugging Face Transformers library. DINO DETR is a state-of-the-art object detection model that builds upon the original DETR architecture, incorporating improvements such as:

- **Contrastive denoising training** to enhance object queries.
- **Mixed query selection** for more robust matching between predictions and ground truths.
- **Look forward twice (LFT)** mechanism to refine object boxes.
- **Efficient matching and distillation techniques** that stabilize bipartite matching loss.

The model achieves strong performance on COCO test-dev (https://paperswithcode.com/sota/object-detection-on-coco).

Fixes #36205 


## What's included
- Implementation of the DinoDetrModel, and DinoDetrForObjectDetection
- Implementation of DinoDetrImageProcessor

## Resources I've used
- I've based this imlementation on the original repo https://github.com/IDEA-Research/DINO/tree/main and I am testing with the "checkpoint0011_4scale.pth" checkpoint which can be found in https://drive.google.com/drive/folders/1qD5m1NmK0kjE5hh-G17XUX751WsEG-h_?usp=sharing 
- The original repository contains custom cuda kernels for the deformable attention. In their place, I've used a copy of the Deformable Attention implementation found in the Deformable Detr implementation of Transformers https://github.com/huggingface/transformers/blob/main/src/transformers/models/deformable_detr/modeling_deformable_detr.py.

## Who can review?
@amyeroberts, @qubvel

Commit:
First commit for DINO - DETR. Copy-paste files from Deformable DETR and original DINO repo. Substitute Deformable for DINO where needed.

Commit:
Add dino detr loss

Commit:
First implementation of dino detr post processor

Commit:
Add misc functions to dino modeling, make config first version

Commit:
Restructure dino detr modeling code, major changes are now required only in the DINO class

Commit:
Match config to source code in modeling dino detr, defaults in config are still missing

Commit:
Clean up dino_detr_config

Commit:
Fixed import errors, code can now be executed

Commit:
Model now can be instantiated with random weights

Commit:
Forward pass for randomly initialized model now executes without errors

Commit:
Forward pass works for DINO model with pretrained weights and matches original repo

Commit:
Add dino detr conversion to pytorch script

Commit:
Ensure model outputs for single COCO image are within 1e-3 error of reference

Commit:
Add draft unit tests

Commit:
Apply make style

konstantinos-p commented:
Hello @qubvel! The state of the PR is that I've gotten the forward pass to match the original implementation up to the required precision. I've marked this as a draft because I wanted to just get a first opinion on if I'm modifying the correct files in the codebase. Let me know if I'm missing anything big. Regarding tests, i've copied some from Deformable Detr but haven't tried getting them to work. Let me know if I need to add any apart from what's already there.

qubvel commented:
Hi @konstantinos-p! Thanks a lot for working on the model, super excited to see it merged! 🚀  

Before diving into the implementation details, here are some general comments we need to address in the PR:  

- [ ] **Integration Tests**  
   Let's set up integration tests to ensure output consistency is maintained while refactoring the modeling code. Please use `torch.testing.assert_close` when comparing tensors (**not use** `self.assertTrue(torch.allclose(...)`)

- [ ] **Modeling Code**  
   It would be super nice to use `modular` approach! It's the way to use inheritance in transformers while keeping a one-model-one-file format, please see more information here: https://huggingface.co/docs/transformers/main/en/modular_transformers and in `examples/modular-transformers` folder of the repo. Also, Siglip2 and RT-DETRv2 model were added using modular.

- [ ] **Consistent Naming for Classes & Modules**  
   - Model names and types: `DinoDetr`  and `dino_detr`
   - All module names should be prefixed with the model name. For example, instead of `ResidualBlock`, use `DinoDetrResidualBlock`.  

- [ ] **Code Paths & Cleanup**  
   - No asserts in code and minimum `raise` statements, config params should be validated in config.
   - Remove unused conditional statements if no existing configuration requires them. Also, remove the unnecessary config parameters.

     ```python
     # Instead of:
     if config.use_last_layer_norm:
         self.norm = nn.LayerNorm()
     else:
         self.norm = nn.Identity()
     ```
     ```python
     # If `use_last_layer_norm = True` in all configs, simplify it to:  
     self.norm = nn.LayerNorm()
     ```

- [ ] **Code Style**  
   - Use clear, descriptive variable names: avoid single-letter (`x`) or placeholder (`tmp`) variables.  
   - Add comments for non-obvious code.  
   - Use type hints for modules (e.g. in `__init__` and `forward`).  

- [ ] **Conversion Script Standard**  
   Please follow the `mllama` model format for the conversion script. Define a single key mapping using regex (see `rt_detr_v2`, `ijepa`, `mllama`, `superglue` models for reference). This is our new standard for all newly added models.  

Thanks again! Looking forward to the updates 🚀

konstantinos-p commented:
Thanks for the comments! I'll start addressing them!

Commit:
Modeling integration test passes

Commit:
Add integration and unit tests

The integration tests and most unit tests are passing. The unit tests
that are failing are mainly due to gradient checkpointing not being supported and due to using shared tensors in the
implementation (which causes saving and loading tests to fail).

Commit:
Remove unnecessary dependency on pretrained model class from some classes

konstantinos-p commented:
@qubvel I'm facing a similar issue as [here](https://github.com/huggingface/transformers/pull/36895#issuecomment-2752547768). Basically I'm rewriting the model to use modular but instead of DinoDetr Dino get's captured by the compiler as the model name. Is there a fix for this? Alternatively I could apply modular only on the processor.

Commit:
All image processing tests pass

Commit:
Consistently start module names with DinoDetr

Commit:
Pass test_hidden_states_output test

Commit:
Dino detr doesn't support gradient checkpointing

Commit:
Pass test_model_is_small test

Commit:
Pass test_model_outputs_equivalence test

Cyrilvallez commented:
Yes, see my answer here https://github.com/huggingface/transformers/pull/36895#issuecomment-2815598690!

Commit:
Pass test_attention_outputs

Commit:
Pass test_retain_grad_hidden_states_attentions

Commit:
Pass safetensor tests

Commit:
All tests pass for the first model draft

All intergration tests (test_modeling_dino_detr.py and test_processing_dino_detr.py) pass
for the current model implementation. This model implementation is a first draft that needs
to be refactored for improved structure and readability.

Commit:
Convert weights to hf using a single key mapping and regex

Commit:
Add some missing copied_from comments, these work with the huggingface compiler

Commit:
Move all asserts to the config

Commit:
Remove some unused config parameters

Commit:
Remove unused config parameters, simplify inits

Commit:
Clean up decoder code

Commit:
Clean up deformable transformer code

Commit:
Clean DinoDetrModel code

Commit:
Clean up DinoDetrForObjectDetection class

Commit:
Add type hints

Commit:
Add a model checkpoint to hub

Commit:
Fix copied from in image_processing_dino_detr

Commit:
Some changes to make copied from mechanism work

Commit:
Fix more code quality errors

Make DinoDetrConvEncoder identical with the original version from the original detr folder. Remove
the copied from comment from the DinoDetrPreTrainedModel class. The original class contains a
gradient_checkpointing=True flag which causes tests to fail. Gradient checkpointing is not implemented
in the current model.

Commit:
Fill missing values config docstring

Commit:
Add dino detr to models/__init__.py file

Commit:
DinoDetrDeformableTransformer is recognized as a model, attempt a fix

Commit:
Attempt fix for DinoDetrDeformableTransformer getting classified as a model

Commit:
Add docs template

Commit:
Add missing model type to the  constant

Commit:
Remove config from loss function definition

Commit:
Remove unused config parameters

Commit:
DinoDetr docs first complete draft

Commit:
Clean docstrings

Commit:
Add docs to toctree.yml

Commit:
Fix some errors after rebase

Commit:
Update deformable attention definition

konstantinos-p commented:
@Cyrilvallez thanks for the fix, it seems to work for me! Some other issues that I've ran into are that if I run `make style` import errors are detected in the modular file even though they are correctly handled by the compiler. This can start limiting the compiler usability as I need to copy different classes inside modular.py to stop the errors.

Commit:
Implement using modular

Commit:
Apply conditions on some imports

Commit:
Remove asserts from postprocessing method

Commit:
Sort imports

konstantinos-p commented:
Hello @qubvel I've addressed your initial comments, I've added tests which pass + a first draft of docs etc. I think it would be a good time to have a second look! CircleCI tests were passing as well but now are failing after I rebased on recent changes. I'll need some help with these as well!

konstantinos-p commented:
Hello @qubvel! It would be great if we could make a final push and merge this! I'd rather work on it while the key points are still fresh. 

DINO Detr is still very competitive and an important model in the Detr line on top of which one could create new architectures.

molbap reviewed (commented):
Did an initial review and tried to leave pointers for `modular`, when it is not possible to inherit from another model let's try and stay aligned with existing methods! Feel free to ping me when you've had time to update the branch and adress initial comments :hugs:

Commit:
Small fixes

Commit:
Style changes from make fixup

Commit:
Small fixes

Commit:
Address reviewer comments

Commit:
Clean up dino_loss import from detection_loss when possible

Several methods in dino_loss can be imported from detection_loss and are common across different detr models.

Commit:
Clean up configuration_dino_detr.py

Commit:
Clean up convert_dino_detr_to_pytorch.py

Commit:
Remove feature extraction file

Commit:
Address reviewer comments

Commit:
Replace _get_activation_fn with ACT2FN

Commit:
Use get_contrastive_denoising_training_group from rt_detr

Commit:
Reuse deformable_detr image_processor

Commit:
Simplify config

Commit:
Clean up modular code, add comments

Commit:
Cleanup of embeddings

Commit:
Cleanup embeddings

molbap commented:
Hey @konstantinos-p, are you sure you've rebased on origin:main? merge or rebase make no difference here as we'll squash in the end, but the `transformers` version should be v5-dev, not 4.52 (and that one in particular was yanked, i.e. removed from PyPi because it had several critical issues)

taking another look now :)

Commit:
Cleanup: remove layer dropout, two_stage options, ref_token, etc

Commit:
Clean modeling code

Remove several config options that made the code complicated, but were unused. Clean the
EncoderDecoder code by reordering, renaming/removing several variables. Add comments to
make steps clearer.

molbap commented:
To fix conflicts usually I just do (after syncing my fork):

```bash
git checkout main
git pull 
git checkout my_branch_to_update
git merge main
```
then fix conflicts that appeared (test_modeling_common should pop out), among a few.

I did that with your branch and got 3 conflicting files, mostly linked to our removal of support for tensorflow and torchscript + some changes on deformable detr testing. 
<img width="570" height="130" alt="image" src="https://github.com/user-attachments/assets/e94c0102-bd3e-4852-a1f3-3c710f07e38f" />
You can just accept all incoming changes.

Once this is done, I rather suggest you use `fixup`, once back to your branch
```bash
pip install -e .[quality]
make fixup
```
that should help clear out some of the code quality issues. The linter has been updated, a few things have changed regarding typing (list, not List), etc, so doing that should clear up the code and the CI a bit :hugs:

Commit:
Add DinoDetrImageProcessor to __all__

Commit:
Make fix copies

Commit:
Revert changes to test modeling deformable detr

Commit:
Small fix

Commit:
Fix docstring

Commit:
Ignore docstring consistency, it throws misleading errors

Commit:
Add architecture link to config docstring

konstantinos-p commented:
All tests are passing on my side! test_modelling_common just needs to be updated to the latest main changes before merging so that some lines have the correct length.

molbap commented on src/transformers/models/dino_detr/modular_dino_detr.py:
```diff
@@ -0,0 +1,2177 @@
+
+if is_timm_available():
+    pass
+
```
```suggestion
```

molbap commented on src/transformers/models/dino_detr/modular_dino_detr.py:
```diff
@@ -0,0 +1,2177 @@
+    return torch.log(x1 / x2)
+
+
+def dn_post_process(
```
nit on naming, "denoising_post_process" is more readable (we can spare the characters)

molbap commented on src/transformers/models/dino_detr/modular_dino_detr.py:
```diff
@@ -0,0 +1,2177 @@
+    else:
+        raise ValueError(f"not supported {config.position_embedding}")
+
+    return position_embeddings
```
On the canonical model that goes with this, which option is selected? We try to have one model = one file, this is a bit of a core change :eyes:

konstantinos-p replied:
Default is SineHW so I'll just remove these options and clean the config

molbap commented on src/transformers/models/dino_detr/modular_dino_detr.py:
````diff
@@ -0,0 +1,2177 @@
+        Detected <object> with confidence 0.7229 at location <location>
+        ```"""
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
````
Likewise, we don't add `output_attentions` and `output_hidden_states` anymore, they are handled through decorators that trace what is needed https://github.com/huggingface/transformers/blob/20396951afaf35e12979e8ef54b2289936921ac6/src/transformers/modeling_utils.py#L1742 and `check_model_inputs` https://github.com/huggingface/transformers/blob/20396951afaf35e12979e8ef54b2289936921ac6/src/transformers/utils/generic.py#L784 , they help make the code much leaner!

molbap reviewed (commented):
Nice for the tests! Left a review to align with the new, leaner API for tuples, attentions and hidden states

molbap commented:
Thanks a lot! Ping me if you encounter issues :hugs:

Commit:
Replace return_dict with can_return_tuple

Commit:
Use _can_record_outputs instead of output_attentions output_hidden_states

Commit:
Simplify code

konstantinos-p commented on src/transformers/models/dino_detr/modular_dino_detr.py:
```diff
@@ -0,0 +1,2177 @@
+    if layer_share:
+        return nn.ModuleList([module for i in range(N)])
+    else:
+        return nn.ModuleList([copy.deepcopy(module) for i in range(N)])
```
I removed all options from the config and hardcoded the choices, in particular the logistic regression layers for predicting classes etc are shared, while layers themselves are not shared.

Commit:
Simplify layer share

Commit:
Make code simpler

molbap commented:
Hey @konstantinos-p , we are currently patching 4.57 and also doing a lot of work to prepare for v5, apologies bandwidth is limited :bow: 
Did you manage to solve you `can_record_outputs` problem?

Commit:
Rename cardinality loss to cardinality error

Commit:
Remove redundant imports from modular

Commit:
Improve docs

Commit:
Fix failing tests after rebase

Cyrilvallez commented:
Sorry about that! Could you rebase/merge with latest main and try again? Tied_weights have been fixed earlier today!

Commit:
Fix most tests apart from test_can_use_safetensors and test_load_save_without_tied_weights

Commit:
Remove tied weights

Cyrilvallez reviewed (commented):
Hey! Here are a few comments!
Also, look like the processor etc were not added to auto mappings!
Finally, it would be much better to break from the `load_backbone` legacy here to add the backbone into the model. Instead, just use `AutoModel`, with the associated config, or if backbone is only `timm`, can be directly `TimmWrapper` for example

Commit:
Simplify code

Commit:
Simplify _can_record_outputs

Commit:
Add auto_docstring

Commit:
Simplify EmbeddingSineHW

Commit:
Simplify code

Commit:
Simplify decoder layer

Commit:
Remove get_clones completely

Commit:
Fix auto_docstring

Commit:
Fix ini_weights

praneet0017 commented:
@konstantinos-p @Cyrilvallez any update on this? :) hoping this model gets added to HF

Cyrilvallez commented:
Hey @praneet0017! Sorry, you need to rebase and fix conflicts before we can review again!
